dozer-ingestion/src/connectors/postgres/connection/client.rs

```diff
@@ -0,0 +1,260 @@
+    client: Arc<Mutex<Client>>,
+    query: String,
+    query_params: Vec<String>,
+    cursor_position: u64,
+    inner: Pin<Box<tokio_postgres::RowStream>>,
+    pending_resume: Option<
+        BoxFuture<'static, Result<(Client, tokio_postgres::RowStream), tokio_postgres::Error>>,
+    >,
+}
+
+impl RowStream {
+    pub async fn new(
+        mut client: Client,
+        query: String,
+        params: Vec<String>,
+    ) -> Result<Self, tokio_postgres::Error> {
+        let statement = client.prepare(&query).await?;
+        let inner = client.query_raw_internal(statement, params.clone()).await?;
+        Ok(Self {
+            client: Arc::new(Mutex::new(client)),
+            query,
+            query_params: params,
+            cursor_position: 0,
+            inner: Box::pin(inner),
+            pending_resume: None,
+        })
+    }
+
+    fn resume(
+        &mut self,
+    ) -> BoxFuture<'static, Result<(Client, tokio_postgres::RowStream), tokio_postgres::Error>>
+    {
+        async fn resume_async(
+            client: Arc<Mutex<Client>>,
+            query: String,
+            params: Vec<String>,
+            offset: u64,
+        ) -> Result<(Client, tokio_postgres::RowStream), tokio_postgres::Error> {
+            let config = client.lock().await.config().clone();
+            // reconnect
+            let mut client = Client::connect(config).await?;
+            // send query with offset
+            let statement = client.prepare(&add_query_offset(&query, offset)).await?;
+            let row_stream = client.query_raw_internal(statement, params).await?;
+            Ok((client, row_stream))
+        }
+
+        Box::pin(resume_async(
+            self.client.clone(),
+            self.query.clone(),
+            self.query_params.clone(),
+            self.cursor_position,
+        ))
+    }
+}
+
+impl Stream for RowStream {
+    type Item = Result<Row, tokio_postgres::Error>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        loop {
+            if let Some(resume) = this.pending_resume.as_mut() {
+                match ready!(resume.as_mut().poll(cx)) {
+                    Ok((client, inner)) => {
+                        this.pending_resume = None;
+                        this.client = Arc::new(Mutex::new(client));
+                        this.inner = Box::pin(inner);
+                    }
+                    Err(err) => return Poll::Ready(Some(Err(err))),
+                }
+            }
+
+            match ready!(this.inner.as_mut().poll_next(cx)) {
+                Some(Ok(row)) => {
+                    this.cursor_position += 1;
+                    return Poll::Ready(Some(Ok(row)));
+                }
+                Some(Err(err)) => {
+                    if is_network_failure(&err) {
+                        this.pending_resume = Some(this.resume());
```

Comment:
Once we resume, this is a branch new stream and will generate all rows again?

I see `cursor_position` which looks like for dedup, but it's not used?

Reply:
It's used here
https://github.com/getdozer/dozer/blob/07094fd6e9cf7749d00937bf210ff0604e6eaeb8/dozer-ingestion/src/connectors/postgres/connection/client.rs#L203
https://github.com/getdozer/dozer/blob/07094fd6e9cf7749d00937bf210ff0604e6eaeb8/dozer-ingestion/src/connectors/postgres/connection/client.rs#L194
The query isn't restarted as is, the offset `cursor_position` is appended to the SQL query.

Reply:
Oh cool. Missed that.